tiledb/sm/filesystem/gcs.h

```diff
@@ -599,6 +620,18 @@ class GCS {
    */
   Status is_bucket(const std::string& bucket_name, bool* const is_bucket) const;
 
+  /**
+   * Contains the implementation of ls_filtered.
+   *
+   * @param uri The parent path to list sub-paths.
+   * @param f The FilePredicate to invoke on each object for filtering.
+   * @param recursive Whether to recursively list subdirectories.
+   */
+  LsObjects ls_filtered_impl(
+      const URI& uri,
+      std::function<bool(const std::string_view, uint64_t)> file_filter,
```

Comment:
This is intentional. If `FileFilter` was generic, the implementation of `ls_filtered_impl` would need to be on `gcs.h`, which would bring `google/cloud/storage/client.h` on the header, which would significantly regress build times as observed in #4777, which I would strongly prefer not to do.

There might be ways to avoid the conversion to `std::function` and still prevent the GCS headers from leaking (like I did with Azure), but since the structure of `ls_recursive` in GCS is different from Azure, trying to come with such solution will not be trivial, and it would definitely be more complicated than the current one.

There has been an argument that the callbacks should be generic all the way in because that's what C++'s `<algorithm>` does, but I do not believe the circumstances are comparable to this case. For starters, the functions in `<algorithm>` are self-contained and relatively trivial, do not perform HTTP requests, and do not depend on third-party components. Furthermore, when/if in the future we decide to refactor VFS to be polymorphic and not directly depend on the specific VFSes, we will be forced to use a non-generic implementation of callbacks.

Reply:
I was suggested offline to:

* make this function generic
* change the implementation of the public-facing `ls_filtered` to pass a lambda that captures the user-provided `file_filter` and calls it
*
  add in `gcs.cc` an explicit template instantiation of `ls_filtered_impl` with the lambda of the previous bullet point like this:

  ```cpp
  using file_predicate_lambda =
    decltype([](const std::string_view, uint64_t) { return true; });

  template LsObjects GCS::ls_filtered_impl<file_predicate_lambda>(
    const URI& uri, file_predicate_lambda file_filter, bool recursive) const;
  ```

However it failed with the following linker error:

```
[build] vfs_api.cc.obj : error LNK2019: unresolved external symbol "private: class std::vector<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64>,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64> > > __cdecl tiledb::sm::GCS::ls_filtered_impl<class `public: class std::vector<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64>,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64> > > __cdecl tiledb::sm::GCS::ls_filtered<class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)> >(class tiledb::sm::URI const &,class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)>,bool)const '::`2'::<lambda_1> >(class tiledb::sm::URI const &,class `public: class std::vector<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64>,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64> > > __cdecl tiledb::sm::GCS::ls_filtered<class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)> >(class tiledb::sm::URI const &,class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)>,bool)const '::`2'::<lambda_1>,bool)const " (??$ls_filtered_impl@V<lambda_1>@?1???$ls_filtered@VCallbackWrapperCAPI@sm@tiledb@@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@std@@@GCS@sm@tiledb@@QEBA?AV?$vector@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@V?$allocator@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@@2@@std@@AEBVURI@34@VCallbackWrapperCAPI@34@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@6@_N@Z@@GCS@sm@tiledb@@AEBA?AV?$vector@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@V?$allocator@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@@2@@std@@AEBVURI@12@V<lambda_1>@?1???$ls_filtered@VCallbackWrapperCAPI@sm@tiledb@@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@std@@@012@QEBA?AV34@0VCallbackWrapperCAPI@12@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@4@_N@Z@3@Z) referenced in function "public: class std::vector<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64>,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,unsigned __int64> > > __cdecl tiledb::sm::GCS::ls_filtered<class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)> >(class tiledb::sm::URI const &,class tiledb::sm::CallbackWrapperCAPI,class std::function<bool __cdecl(class std::basic_string_view<char,struct std::char_traits<char> > const &)>,bool)const " (??$ls_filtered@VCallbackWrapperCAPI@sm@tiledb@@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@std@@@GCS@sm@tiledb@@QEBA?AV?$vector@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@V?$allocator@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@std@@@2@@std@@AEBVURI@12@VCallbackWrapperCAPI@12@V?$function@$$A6A_NAEBV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z@4@_N@Z)
```

Apparently the types of captured values are part of the lambda's type, and combined with [some further research](https://stackoverflow.com/a/4846597), the proposed alternative design seems impossible. I still don't see a problem with the present design of using `std::function`. Sure you can't capture certain types by value but that's an edge-case scenario that is not currently affecting us, and would be resolved by using `std::move_only_function` when we switch to C++ 23. I will add documentation mentioning all that.

c.c. @KiterLuc @eric-hughes-tiledb

Reply:
Updated documentation.

Reply:
Looking into this.